src/plugins/discover/public/application/main/hooks/utils/build_state_subscribe.ts

```diff
@@ -43,22 +44,45 @@ export const buildStateSubscribe =
     setDataView: DiscoverStateContainer['actions']['setDataView'];
   }) =>
   async (nextState: DiscoverAppState) => {
+    let savedSearchDataView;
     const prevState = appState.getPrevious();
     const savedSearch = savedSearchState.getState();
     if (isEqualState(prevState, nextState)) {
       addLog('[appstate] subscribe update ignored due to no changes', { prevState, nextState });
       return;
     }
     addLog('[appstate] subscribe triggered', nextState);
-    const { hideChart, interval, breakdownField, sort, index } = appState.getPrevious();
+    const { hideChart, interval, breakdownField, sort } = appState.getPrevious();
+    const nextQuery = nextState.query;
+    const currentDataView = internalState.getState().dataView;
+
+    if (isTextBasedQuery(nextQuery)) {
+      const nextDataView = await getDataViewByTextBasedQueryLang(
+        nextQuery,
+        currentDataView,
+        services
+      );
+      if (currentDataView !== nextDataView) {
+        addLog('[appstate] text based language change of data view', {
+          current: currentDataView,
+          next: nextDataView,
+          query: nextQuery,
+        });
+        setDataView(nextDataView);
+        savedSearchDataView = nextDataView;
+      }
+    }
```

Comment:
FYI @davismcphee, nothing for this PR. For the text based query language, and when all aka `*` fields are requested, we would now have the potential columns by using the data view's field available.